RFC: Arch: Basic MIPS port.

IMGJulian opened the pull request:
Supports the MIPS Danube SOC and the Creator Ci40 board.
Runs in single vector mode using the IRQ table.
Kernel mode only, no direct IRQ support.
Compiled using the ImgTec mips-mti-elf toolchain.

Signed-off-by: julian ingram <julian.ingram@imgtec.com>

SebastianBoe commented on Makefile:
```diff
@@ -403,9 +403,9 @@ KBUILD_CXXFLAGS   := -c -g -std=c++11 \
 KBUILD_AFLAGS   := -c -g -xassembler-with-cpp -D_ASMLANGUAGE
 
-LDFLAGS += $(call ld-option,-nostartfiles)
+LDFLAGS += $(if $(filter $(ARCH), mips),,$(call ld-option,-nostartfiles))
```
Could there be an explanation here in english explaining what is special about mips wrt. -nostartfiles ?

IMGJulian replied:
Currently this port uses `_init` and `_fini`, I realise you would like this to be more self contained, we are looking into how best to do this.

nashif commented:
You will want to split this into arch, soc, board, driver commits, it would also nice to have a qemu port to test this without having HW.

SebastianBoe commented on arch/mips/soc/danube/Makefile:
```diff
@@ -0,0 +1,17 @@
+soc-cflags := -I/$(srctree)/arch/$(ARCH)/soc/$(SOC_PATH)/ -I$(GCCMIPSMTI_TOOLCHAIN_PATH)/mips-mti-elf/include -EL -mips32r2
```
Could you set some variable in the Makefile.toolchain* file so that you don't have to hardcode mips-mti-elf ?

Then this SOC wouldn't depend on a mips toolchain with this exact signature.

Makefile.toolchain.zephyr sets SYSROOT_TARGET_${ARCH} for this purpose.

Also, shouldn't the Makefile.toolchain.* file be putting this directory on the include path, and not the SoC?

andrewboie commented on Makefile:
```diff
@@ -403,9 +403,9 @@ KBUILD_CXXFLAGS   := -c -g -std=c++11 \
+LDFLAGS += $(if $(filter $(ARCH), mips),,$(call ld-option,-nostartfiles))
 LDFLAGS += $(call ld-option,-nodefaultlibs)
-LDFLAGS += $(call ld-option,-nostdlib)
+LDFLAGS += $(if $(filter $(ARCH), mips),,$(call ld-option,-nostdlib))
```
Same as above comment by Sebastian, why do we need this? What symbols are missing without it? We explicitly include libgcc later on in the build IIRC

andrewboie commented on Makefile:
```diff
@@ -850,7 +850,7 @@ quiet_cmd_create-lnk = LINK    $@
 	echo "-L $(objtree)/include/generated";					\
 	echo "-u _OffsetAbsSyms -u _ConfigAbsSyms"; 				\
-	echo "-e __start"; 						 	\
+	echo "$(if $(filter $(ARCH), mips),,-e __start)"; 			\
```
Please also explain this. Why do we need different entry points in some cases?

andrewboie commented on arch/mips/core/isr.S:
```diff
@@ -0,0 +1,251 @@
+    move    t4, zero
+    move	t5, zero
+    mfc0	t2, C0_CONFIG3
+    ext	    t3, t2, CFG3_BP_SHIFT, 1
```
please use consistent spacing between ASM mnemonics and their parameters, it is all over the place in this and other asm files. on other arches we typically use a tab.

andrewboie commented on arch/mips/core/prep_c.c:
```diff
@@ -0,0 +1,45 @@
+void software_init_hook(void)
+{
+#ifdef CONFIG_XIP
+	memcpy(&_fdata, &_rom_data_copy, (int)&_edata -(int)&_fdata);
```
please setup the proper linker symbols and use _data_copy() in kernel/init.c

andrewboie commented on arch/mips/core/swap.S:
```diff
@@ -0,0 +1,64 @@
+SECTION_FUNC(TEXT, _thread_entry_wrapper)
+	/*
+	 * _thread_entry_wrapper is called for every new thread upon the return
+	 * of __swap or ISR. Its address, as well as its input function
```
This comment seems misleading.
_thread_entry_wrapper is *only* called upon the initial __swap() to a new thread.

andrewboie commented on arch/mips/core/swap.S:
```diff
@@ -0,0 +1,64 @@
+	 * _thread_entry_wrapper is called for every new thread upon the return
+	 * of __swap or ISR. Its address, as well as its input function
+	 * arguments thread_entry_t, void *, void *, void * are restored from
+	 * the thread stack (initialized via function _thread).
```
initalized via function _new_thread()

andrewboie commented on arch/mips/core/swap.S:
```diff
@@ -0,0 +1,64 @@
+	 * to function _thread_entry. Hence, just call _thread_entry with
+	 * return address set to 0 to indicate a non-returning function call.
+	 */
+
```
What mechanism pops the initial arguments off the stack? This function is a no-op. If this is all done transparently by the calling convention, why do we need this at all?

andrewboie commented on arch/mips/core/thread.c:
```diff
@@ -0,0 +1,55 @@
+		 */
+		.status = (reg_t)0x20008103,
+		.cause = mips_getcr() & ~(CR_IP_MASK),
+		.epc = (reg_t)_thread_entry_wrapper
```
As mentioned earlier, since _thread_entry_wrapper doesn't seem to do much at all, can we just directly put _thread_entry here instead of _thread_entry_wrapper?

andrewboie commented on arch/mips/core/thread.c:
```diff
@@ -0,0 +1,55 @@
+
+	thread->callee_saved.sp = (u32_t)ctx;
+
+	struct gpctx stack_init = {
```
i don't like the memcpy this requires, it seems unnecessary.
Why not cast &stack_init to a struct gpctx * and initialize directly, after the call to _new_thread_init()?

IMGJulian replied:
I agree, but would that not be a strict aliasing violation in cases where the stack is not made up of chars?

andrewboie replied:
not sure what you mean here.
K_THREAD_STACK_BUFFER() return type is always char *.

IMGJulian replied:
I assumed you meant `ctx` should be cast to `struct gpctx *` and then initialised (as `&stack_init` is already a `struct gpctx *`). If so, I think that violates strict aliasing b/c `ctx` is just an offset from `stack` which is of the potentially incompatible type `k_thread_stack_t`. I am not calling `K_THREAD_STACK_BUFFER()` to initialise `ctx`, but even if I was, I think my point still stands (but I could be totally wrong)...

andrewboie replied:
Your code is wrong above, You should get the underlying buffer of the stacl with K_THREAD_STACK_BUFFER()

andrewboie replied:
have a look at the x86 port of _new_thread() to see what I am looking for.

IMGJulian replied:
You're right, I will change that. You haven't quite convinced me to get rid of the `memcpy` yet though!

andrewboie replied:
K_THREAD_STACK_BUFFER() returns a char *. Do the appropriate math on it and cast your data structure. Why is this a problem? Please, get rid of memcpy(). It casts its arguments to char* anyway. If we are unable to treat stack buffers at character buffers the whole OS won't work.

IMGJulian replied:
I'm not trying to be difficult here. If it is guaranteed that the stack buffer will always be a char type (not just that `K_THREAD_STACK_BUFFER()` will always return a `char *` type) then that is fine. This seems to be the case from your comment below.

andrewboie commented on arch/mips/include/kernel_arch_func.h:
```diff
@@ -0,0 +1,81 @@
+	/* the aligned attribute here is the only difference between
+	 * this and the standard switch
+	 */
+	struct k_thread dummy_thread_memory __aligned(8);
```
what's special about the dummy thread that it has to be aligned(8), unlike all other struct k_thread in the system?

andrewboie commented on arch/mips/core/fatal.c:
```diff
@@ -0,0 +1,72 @@
+ *
+ * @return This function does not return.
+ */
+FUNC_NORETURN void _NanoFatalErrorHandler(unsigned int reason,
```
this isn't an acceptable implementation.
_NanoFatalErrorHandler prints an error message and then calls _SysFatalErrorHandler, which implements policy on what to do when a thread encounters an exception.

andrewboie commented on arch/mips/core/fatal.c:
```diff
@@ -0,0 +1,72 @@
+ *
+ * @return N/A
+ */
+FUNC_NORETURN __weak void _SysFatalErrorHandler(unsigned int reason,
```
Please see how this is done on other arches, at minimum we need logic to decide whether to just abort the current thread or halt the system.

andrewboie commented on arch/mips/core/fatal.c:
```diff
@@ -0,0 +1,72 @@
+
+FUNC_NORETURN void _Fault(const NANO_ESF *esf)
+{
+	_NanoFatalErrorHandler(_NANO_ERR_CPU_EXCEPTION, esf);
```
It would be really good to identify what specific type of CPU exception happened here and dump out any interesting information about it.

andrewboie commented on arch/mips/core/irq_manage.c:
```diff
@@ -0,0 +1,54 @@
+void _irq_spurious(void *unused)
+{
+	ARG_UNUSED(unused);
+	while (1) {
```
don't do this....need to go through _NanoFatalErrorHandler path

andrewboie commented on arch/mips/soc/danube/danube_irq.c:
```diff
@@ -0,0 +1,65 @@
+	int key;
+
+	key = irq_lock();
+	mips_bissr(1 << (irq + 8));
```
where is mips_bissr (and other related functions) actually implemented?
is it some kind of runtime library?
we would prefer zephyr to be stand-alone, if these are just functions to poke registers can you port them directly to zephyr?

andrewboie commented on include/arch/mips/danube/linker.ld:
```diff
@@ -0,0 +1,352 @@
+     IDT_LIST  (wx)      : ORIGIN = 0xFFFFF7FF, LENGTH = 2K
+ }
+
+__entry = DEFINED(__reset_vector) ? MIPS_ROM_BASE_ADDR : _start;
```
what's the story with this? why do we need to conditionally define the entry point?

tyrantlizard77 reviewed (changes requested):
Some initial review impressions.

Please understand that in order to accept this into the open source project, we are going to need to see this working with a MIPS build of GCC, and demonstrate our test cases passing with sanitycheck using a MIPS build of QEMU. You can have a toolchain definition for a proprietary compiler, but support for some free open source compiler and emulator that we can bundle into the SDK is a hard requirement, primarily so we can assert that the port doesn't bit-rot over time by having sanitycheck builds verify it on an ongoing basis in our CI infrastructure.

I would get with Anas and Ram and see about getting MIPS tools (compiler and QEMU) included in the Zephyr SDK.

It seems this arch port depends on some external HAL code. How heavyweight is this HAL?

IMGJulian commented:
Thanks for all your comments, it seems MIPS QEMU support would be useful, I will see what I can do.

To be clear, the ImgTec compiler is open source and GCC-based, the compiler and HAL can be found here: https://community.imgtec.com/developers/mips/tools/codescape-mips-sdk/

andrewboie commented:
We could consider then including the ImgTec compiler in the SDK if the licensing terms are agreeable.

andrewboie commented on arch/mips/core/thread.c:
```diff
@@ -0,0 +1,55 @@
+	_ASSERT_VALID_PRIO(prio, entry);
+
+	char *stack_memory = K_THREAD_STACK_BUFFER(stack);
+	char *ctx = (char *)STACK_ROUND_DOWN(stack + stack_size -
```
This is wrong, please initialize ctx in terms of stack_memory not stack.
k_thread_stack_t is an opaque type, used for inserting guard regions for memory protection scenarios.

nashif commented:
> We could consider then including the ImgTec compiler in the SDK if the licensing terms are agreeable.

First I would like to try the existing MIPS support in our SDK which was disabled, building it now, we will see how that goes, it also enabled qemu support for mips

galak reviewed (dismissed):
I would request any new arch/board port use DTS from the start.

nashif commented:
can this be rebased on top of master please?

IMGJulian commented:
Yes. BTW, the MIPS toolchain in the Zephyr SDK only seems to support big endian.

IMGJulian commented:
There are quite a few files from the MIPS HAL that do not share your coding style (CMSIS seems similar) can I safely ignore the checkpatch errors for these?

nashif commented:
> There are quite a few files from the MIPS HAL that do not share your coding style (CMSIS seems similar) can I safely ignore the checkpatch errors for these?

those will need to go into ext/, ext/hal is excluded from checkpatch

Commit:
Arch: Basic MIPS port.

Runs in single vector mode using the IRQ table.
Kernel mode only, no direct IRQ support.
Compiled using the ImgTec mips-mti-elf toolchain.

Signed-off-by: julian ingram <julian.ingram@mips.com>

Commit:
Drivers: MIPS core timer driver.

Implements the system clock tick using the MIPS core timer.
Compiled using the ImgTec mips-mti-elf toolchain.

Signed-off-by: julian ingram <julian.ingram@mips.com>

Commit:
Ext: MIPS HAL.

MIPS HAL from the mips-mti-elf toolchain.

Signed-off-by: julian ingram <julian.ingram@mips.com>

Commit:
SOC: MIPS Danube.

Support for a MIPS InterAptiv core as implemented in the Danube SOC.

Signed-off-by: julian ingram <julian.ingram@mips.com>

Commit:
Boards: Creator Ci40

Skeleton board support for the Ci40 board forom Imagination
Technologies.

Signed-off-by: julian ingram <julian.ingram@mips.com>

Commit:
Boards: MIPS QEMU

Skeleton board support for QEMU MIPS simulation.

Signed-off-by: julian ingram <julian.ingram@mips.com>

IMGJulian commented:
> those will need to go into ext/, ext/hal is excluded from checkpatch
Thanks, that did the trick,

Any chance you think this is ready to become a regular PR? I think I have addressed most of the comments above.

nashif commented on boards/mips/qemu_mipsel/qemu_mipsel.yml:
```diff
@@ -0,0 +1,7 @@
+identifier: qemu_mipsel
```
rename this file to be .yaml, we will support yml in the scripts, but right now .yml is being ignored. Once you do that, you will have lots of sanitycheck errors, so we need to fix those.

IMGJulian replied:
OK fixing that now. I think my toolchain must be picking up mips/cpu from its install dir rather than ext/hal/mips/include/

Last time I checked your MIPS toolchain only supported big-endian targets, so will not compile for the currently written boards/emulators.

IMGJulian replied:
I fixed the sanity check by adding to the enum in the sanity check script. I have a feeling this is not an acceptable solution, so I will probably need to read up on what those yaml files actually do...

nashif replied:
yes, out mips toolchain is not well tested, need to fix this. re the enum, I am thinking now this should not be an enum and should be free form instead to allow anyone to add board/toolchain combos without having to change the scripts.

nashif reviewed (changes requested):
> Any chance you think this is ready to become a regular PR? I think I have addressed most of the comments above.

yes, this looks very good already, however, we still need to be able to build many of the tests using qemu, this is not working now. I just tried with a few hacks to get this building and hit issues , for example:

```
-o CMakeFiles/offsets.dir/arch/mips/core/offsets/offsets.c.obj   -c /work/zephyr/github/zephyr/arch/mips/core/offsets/offsets.c
In file included from /work/zephyr/github/zephyr/include/arch/cpu.h:25:0,
                 from /work/zephyr/github/zephyr/include/kernel.h:33,
                 from /work/zephyr/github/zephyr/kernel/include/kernel_structs.h:10,
                 from /work/zephyr/github/zephyr/arch/mips/core/offsets/offsets.c:2:
/work/zephyr/github/zephyr/include/arch/mips/arch.h:21:22: fatal error: mips/cpu.h: No such file or directory
 #include <mips/cpu.h>
```

Will the zephyr SDK be able to build the qemu target? what are we missing?

galak commented:
> Thanks, that did the trick,
> 
> Any chance you think this is ready to become a regular PR? I think I have addressed most of the comments above.

I think you can remove the RFC, but there are several issues that need to be resolved before its ready to merge.

1. Can you update the commit message for the import into ext/ per CONTRIBUTING.rst (see towards the bottom)
2. Can you provide a simple uart driver - not sure if that is one or two based on what qemu is emulating
3. As per @nashif comments we should be get this building with the Zephyr SDK MIPS toolchain and either add what support we need to the SDK to make that usable.

nashif commented on include/arch/mips/danube/linker.ld:
```diff
@@ -0,0 +1,383 @@
+
+OUTPUT_FORMAT("elf32-tradlittlemips", "elf32-tradbigmips", \
+    "elf32-tradlittlemips")
+GROUP(-lc -luhi -lgcc)
```
I am trying with a toolchain built with xtools, everything builds fine but at the final linking stage I hit this, first the output format is unknown (changing to elf32-littlemips works) and then I can't link with -luhi. What is libuhi and and do we need this with the hal in ext/?

nashif commented:
I also have issues with downloaded toolchain (Codescape.GNU.Tools.Package.2017.10-05.for.MIPS.IMG.Bare.Metal.CentOS-5.x86_64.tar.gz), can you provide instructions how we can build with this toolchain?

neiljay commented:
> I also have issues with downloaded toolchain (Codescape.GNU.Tools.Package.2017.10-05.for.MIPS.IMG.Bare.Metal.CentOS-5.x86_64.tar.gz)

@nashif, the 'IMG' toolchain is for revision 6 of the MIPS architecture, Julian's port covers r2->r5 which requires this toolchain:

https://codescape.mips.com/components/toolchain/2017.10-05/Codescape.GNU.Tools.Package.2017.10-05.for.MIPS.MTI.Bare.Metal.CentOS-5.x86_64.tar.gz

nashif commented:
stale PR, closing for now.
